### PR TITLE
GLOB-26190 - moving cache code into nodule-graphql

### PR DESCRIPTION
**Pull request by unorthodoxgeek**
Why?
Our FE gateway is where we tested our caching code, and now that proved itself, we
want to reuse it in the rest of our gateways. `nodule-graphql` is where both our
dedup and batching service wrappers reside, so it's where the new code will live as well.

What I've done:
1. Moved caching service wrapper code from our FE gateway
2. Refactored a couple of piece of that code (using default validators as functions, separating logging, etc)
3. Removed references to globality from tests and mocks (using Rick & Morty terms)
4. Wrote a new README item in the services README, explaining the way to turn caching on

### CODE REVIEW

File: src/services/caching/JSONStringify.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/* Stringify an object
+   Comatabile to the BE services swagger (Should return the exact same string)
```
Comment (jessemyers) [inline src/services/caching/JSONStringify.js]: I don't think the audience of this (opensource) library will know what "BE services swagger" means, at least not with the level of context necessary to understand this comment.

Also: spelling `Compatible`

File: package.json [hunk 2]
```diff
@@ -18,12 +18,15 @@
...
         "dataloader": "^1.4.0",
+        "enum": "^2.4.0",
         "express": "^4.16.3",
         "express-jwt": "^5.3.1",
         "helmet": "^3.12.0",
         "http-status-codes": "^1.3.0",
+        "is-uuid": "^1.0.2",
         "jsonwebtoken": "^8.2.0",
         "lodash": "^4.17.5",
+        "node-spookyhash-v2": "^1.0.1",
```
Comment (jessemyers) [inline package.json]: I think the `README` needs some documentation of the `ETag` expectations.

File: src/services/caching/operations.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+/* Store a value at a key with a ttl.
+ *
+ * Uses memached `Add` operation, meaning that the gateway will never overwrite
+ * an existing key. This logic allows backend services to implement invalidation with
+ * by writing a null value with a short TTL (e.g. 5s), thereby avoiding race
+ * conditions where gateway rewrites a stale value back into the cache after
+ * cache invalidation but before reading the new one.
+ */
+// XXX - move to a utility
```
Comment (saivickna) [inline src/services/caching/operations.js]: is this still relevant?

File: src/services/caching/__tests__/cached.test.js [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+            },
+        },
+        logger: {
+            warning: () => {},
+            info: () => {},
+        },
+    }),
+}));
+
+// const noduleConfig = require.requireMock('@globality/nodule-config');
```
Comment (saivickna) [inline src/services/caching/__tests__/cached.test.js]: cleanup?

File: src/services/caching/__tests__/spooky128.test.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import spooky128 from '../spooky128';
+
+describe('spooky', () => {
```
Comment (saivickna) [inline src/services/caching/__tests__/spooky128.test.js]: maybe provide a link to the microcosm repo here?

File: src/services/README.md [hunk 6]
```diff
@@ -83,6 +83,62 @@ to be included during batched requests, the `assignArgs` parameter can be used.
...
+        );
+        return client;
+    }
+    bind('createCacheClient', () => createCacheClient);
+
+app.js:
+
+    import { bindServices, setCache } from '@globality/nodule-graphql';
+    // under bindServices();
+    setCache();
```
Comment (saivickna) [inline src/services/README.md]: you need to have `cache.enabled` in the config set to true for this to work. You should mention that here

File: src/services/caching/logging.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    cacheUsage.keys.push({
+        resourceName: spec.resourceName,
+        args,
+        key,
+        result: result.key,
+        ...extra,
+    });
+}
+
+function calculateExecuteTime(executeStartTime) {
```
Comment (saivickna) [inline src/services/caching/logging.js]: Import from here: https://github.com/globality-corp/nodule-express/blob/c550bd630fea691455b7320f2694973ca3babafa/src/logging/index.js#L7

File: src/services/index.js [hunk 8]
```diff
@@ -4,6 +4,8 @@ import getServiceWrappers from './wrapper';
 import createKeyFunc from './core/keys';
 
 export { default as named } from './core/named';
+export * from './caching/types';
```
Comment (saivickna) [inline src/services/index.js]: I think we like to be explicit about the exports when possible

File: src/services/caching/JSONStringify.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+/* Stringify an object
+   Compatible to microcosm-flask's services swagger (Should return the exact same string)
```
Comment (pelegdvir) [inline src/services/caching/JSONStringify.js]: Thanks

File: src/services/caching/operations.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+/* Caching operations.
```
Comment (pelegdvir) [inline src/services/caching/operations.js]: Can we move those functions to https://github.com/globality-corp/nodule-memcached/pull/1

File: src/services/caching/types.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    validateEtag(req, cacheData) {
+        const requestedEtags = get(req, `cacheControl.etags.${this.resourceName}`);
+        const { config } = getContainer();
+
+        // etag validation not required
+        if (!requestedEtags) {
+            return true;
+        }
+        const { logger } = getContainer();
+        logger.info(req, 'Requested etags', requestedEtags);
```
Comment (pelegdvir) [inline src/services/caching/types.js]: nit: `logger.debug`

File: src/services/caching/types.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        const formttedEtag = `W/"${cachedEtag}"`;
+        const etagMatch = includes(requestedEtags, formttedEtag);
+
+        // if we will see many messages like that, we should consider optimization
+        if (!etagMatch) {
+            if (get(config, 'logger.cache.enabled', false)) {
+                logger.info(req, 'Etag mismatch', { resourceName: this.resourceName });
+            }
+        } else {
+            logger.info(req, 'Etag match', { resourceName: this.resourceName });
```
Comment (pelegdvir) [inline src/services/caching/types.js]: nit: `logger.debug`

File: src/services/core/keys.js [hunk 13]
```diff
@@ -18,15 +19,16 @@ function valueToString(value) {
 
 /* Generate a hashable key for a specific service request.
  *
- * Used for batching, and deduplication.
+ * Used for batching, caching and deduplication.
  *
  */
 const createKey = (args, keyName = '') => {
+    const namespace = get(getContainer('config.cache'), 'namespace', uuidv5.URL);
```
Comment (pelegdvir) [inline src/services/core/keys.js]: nice

File: src/services/caching/wrapper.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        return [serviceData, CacheResult.noop];
+    } catch (err) {
+        const result = CacheResult.error;
+        logger.warning(req, err.message, {
+            error: err,
+            key,
+            args,
+        });
+        trace(spec, req, args, key, result);
```
Comment (pelegdvir) [inline src/services/caching/wrapper.js]: IRL conversation: let's separate `trace` from this function

File: src/services/caching/wrapper.js [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        const res = await wrapped(req, args);
```
Comment (pelegdvir) [inline src/services/caching/wrapper.js]: IRL conversation: can we call `wrapped` twice?

File: src/services/caching/logging.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
+import { get, set } from 'lodash';
+import { getContainer } from '@globality/nodule-config';
+import { extractLoggingProperties } from '@globality/nodule-logging';
+import { calculateExecuteTime } from '@globality/nodule-express';
+/* Trace a cache operation.
+ *
+ * Returned via local cacheUsage extension.
+ */
+export function trace(spec, req, args, key, result, extra) {
```
Comment (pelegdvir) [inline src/services/caching/logging.js]: nit: can we move this function to a separate file (tracing and logging are different)

File: src/services/caching/wrapper.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+ * Using a data loader allows us to dedup and aggregate cache requests,
+ * which is especially useful with memached's `Multi-Get` operation.
+ */
+export function getCacheLoader(req) {
+    // Get a DataLoader for cache operations
+    // Create it if it's the first use of it for the req.
+    const { cache } = getContainer();
+    let loader = get(req, 'loaders.cache');
+    if (!loader) {
+        loader = new DataLoader(argsList => cache.safeGet(req, argsList));
```
Comment (pelegdvir) [inline src/services/caching/wrapper.js]: nice

File: src/services/caching/logging.js (hunk 16 above)
Comment (pelegdvir) [inline src/services/caching/logging.js]: nit: can we rename to `traceCacheCall`

**Pull request merged**